src/lib/mdns/minimal/responders/QueryResponder.h

```diff
@@ -0,0 +1,230 @@
+            mInfo->additionalQName           = qname;
+        }
+        return *this;
+    }
+
+    bool IsValid() const { return mInfo != nullptr; }
+
+private:
+    Internal::QueryResponderInfo * mInfo;
+};
+
+class QueryResponderIterator : public std::iterator<std::input_iterator_tag, QueryResponderRecord>
+{
+public:
+    QueryResponderIterator() : mAdditionalOnly(false), mCurrent(nullptr), mRemaining(0) {}
+    QueryResponderIterator(bool additionalOnly, Internal::QueryResponderInfo * pos, size_t size) :
+        mAdditionalOnly(additionalOnly), mCurrent(pos), mRemaining(size)
+    {
+        SkipInvalid();
+    }
+    QueryResponderIterator(const QueryResponderIterator & other) = default;
+    QueryResponderIterator & operator=(const QueryResponderIterator & other) = default;
+
+    QueryResponderIterator & operator++()
+    {
+        if (mRemaining != 0)
+        {
+            mCurrent++;
+            mRemaining--;
+        }
+        SkipInvalid();
+        return *this;
+    }
+
+    QueryResponderIterator operator++(int)
+    {
+        QueryResponderIterator tmp(*this);
+        operator++();
+        return tmp;
+    }
+
+    bool operator==(const QueryResponderIterator & rhs) const { return mCurrent == rhs.mCurrent; }
+    bool operator!=(const QueryResponderIterator & rhs) const { return mCurrent != rhs.mCurrent; }
+
+    QueryResponderRecord & operator*() { return *mCurrent; }
+    QueryResponderRecord * operator->() { return mCurrent; }
+
+    Internal::QueryResponderInfo * GetInternal() { return mCurrent; }
+    const Internal::QueryResponderInfo * GetInternal() const { return mCurrent; }
+
+private:
+    /// Checks if the value pointed to by 'mCurrent' is a valid reply value.
+    ///
+    /// Valid is being described as
+    ///   - has a responder
+    ///   - record type matches the 'addtional only' setting for the iterator.
+    bool isCurrentValid() const
+    {
+        if (mCurrent->responder == nullptr)
+        {
+            return false;
+        }
+
+        if (mAdditionalOnly && !mCurrent->reportNowAsAdditional)
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    /// Skips invalid/not useful values.
+    /// ensures that if mRemaining is 0, mCurrent is nullptr;
+    void SkipInvalid()
+    {
+        while ((mRemaining > 0) && !isCurrentValid())
+        {
+            mRemaining--;
+            mCurrent++;
+        }
+        if (mRemaining == 0)
+        {
+            mCurrent = nullptr;
+        }
+    }
+
+    bool mAdditionalOnly;
+    Internal::QueryResponderInfo * mCurrent;
+    size_t mRemaining;
+};
+
+/// Responds to mDNS queries.
+///
+/// In particular:
+///   - replies data as provided by the underlying responders
+///   - replies to "_services._dns-sd._udp.local."
+///
+/// Maintains a stateful list of 'additional replies' that can be marked/unmarked
+/// for query processing
+class QueryResponderBase : public Responder // "_services._dns-sd._udp.local"
```

Comment:
It seems really strange that QueryResponderBase is-a Responder and also accepts an array of other responders. What's the reason for using inheritance here? Could the base responder be composed instead of inherited?